internal/server/providers/oidc_test.go

```diff
@@ -2,16 +2,850 @@ package providers
+type tokenResponse struct {
+	code int
+	body string
+}
+
+// testOIDCServer mocks the expected responses from an OIDC provider
+type testOIDCServer struct {
+	userInfoResponse string
+	tokenResponse    tokenResponse
+	signingKey       *rsa.PrivateKey
+
+	knownClientID     string
+	knownClientSecret string
+	knownAuthCode     string // in this test case the auth code can be re-used
+}
+
+const (
+	oktaInvalidClientIDResp = `{
+		"errorCode": "invalid_client",
+		"errorSummary": "Invalid value for 'client_id' parameter.",
+		"errorLink": "invalid_client",
+		"errorId": "aaabbb",
+		"errorCauses": []
+	}`
+
+	//nolint:gosec // this is not an actual secret credential
+	oktaInvalidClientSecretResp = `{
+		"error": "invalid_client",
+		"error_description": "The client secret supplied for a confidential client is invalid."
+	}`
+
+	oktaInvalidAuthCodeResp = `{
+		"error": "invalid_grant",
+		"error_description": "The authorization code is invalid or has expired."
+	}`
+)
+
+func setupDB(t *testing.T) *gorm.DB {
+	driver, err := data.NewSQLiteDriver("file::memory:")
+	assert.NilError(t, err)
+
+	patch.ModelsSymmetricKey(t)
+	db, err := data.NewDB(driver, nil)
+	assert.NilError(t, err)
+
+	return db
+}
+
+func (ts *testOIDCServer) run(t *testing.T) string {
+	newMux := http.NewServeMux()
+	server := httptest.NewTLSServer(newMux)
+
+	// this is the jwks public key that corresponds with _testdata/test-server-sec.key we just loaded
+	jwks := `{
+		"keys": [
+			{
+				"kty": "RSA",
+				"n": "17B6CrEmJYA2-bPUxaL0OqBg3PWs19ab2DIwrgNc2vPdseV2rUWtXTjVFhVM4cfhuJRULpGJHw8Us3YyJNrIWOaPhOlLCOkrtWK3jCl2xEhgLkpyGxLzsjkSGmOmTNtz44dnxW3WNW-3fKr-aAlGI9O81x3tbU7NnNvvFkotC2HvElQIjISw118C13b9SY1Xc7iKJYcwu6NiSnnkazGGkDLZcR0Ja4lp7Iym8sIbQ72o_mGUUDoCO3CDJePliq4RRKvxCPH2SOuzRpUGwzyK4bfZPXRR17GcOd7DJU6kfGxfX6oAuJCqCEUZCYfQYo6Uj4cBl0BPANlPQuJk7irRvw",
+				"e": "AQAB",
+				"alg": "RS256",
+				"kid": "openid-test",
+				"use": "sig"
+			}
+		]
+	}`
+
+	wellKnown := fmt.Sprintf(`{
+		"issuer": "%[1]s",
+		"authorization_endpoint": "%[1]s/auth",
+		"token_endpoint": "%[1]s/token",
+		"jwks_uri": "%[1]s/keys",
+		"userinfo_endpoint": "%[1]s/userinfo",
+		"id_token_signing_alg_values_supported": ["RS256"]
+	}`, server.URL)
+
+	// general OIDC endpoints
+	newMux.HandleFunc("/.well-known/openid-configuration", func(w http.ResponseWriter, req *http.Request) {
+		_, err := io.WriteString(w, wellKnown)
+		if err != nil {
+			w.WriteHeader(500)
+		}
+	})
+	newMux.HandleFunc("/keys", func(w http.ResponseWriter, req *http.Request) {
+		w.Header().Add("Content-Type", "application/json")
+		_, err := io.WriteString(w, jwks)
+		if err != nil {
+			w.WriteHeader(500)
+		}
+	})
+	newMux.HandleFunc("/token", func(w http.ResponseWriter, req *http.Request) {
+		w.Header().Add("Content-Type", "application/json")
+		w.WriteHeader(ts.tokenResponse.code)
+		_, err := io.WriteString(w, ts.tokenResponse.body)
+		if err != nil {
+			w.WriteHeader(500)
+		}
```

Comment:
I don't think this works to report an error. If there was already a status code written (2 lines above) the second call to `WriteHeader` will fail and log an error (https://github.com/golang/go/blob/go1.18.3/src/net/http/server.go#L1140-L1143), but that log won't fail the test for us.

In the other cases where we don't write a header, the call to `io.WriteString`, will write a 200 OK header (https://github.com/golang/go/blob/go1.18.3/src/net/http/server.go#L1600-L1601), so before we get an error from writing the body there will always be a header.

I think instead of trying to write a 500 response, we can fail the test:

```go
assert.Check(t, err, "failed to write response")
```

https://pkg.go.dev/gotest.tools/v3/assert#Check uses `t.Error` so it's safe to call from other goroutines (where as `Assert`, and `t.Fatal` are not, so we shouldn't use those here).